### PR TITLE
Add new parameter to PacketList.sr() [ready for merge]

### PR DESCRIPTION
**Pull request by polybassa**
This parameter limits the lookahead width to search for a answer packet in the remaining list.
With this parameter, the sr() function can be speed up drastically on long PacketLists

### CODE REVIEW

File: scapy/plist.py [hunk 1]
```diff
@@ -700,7 +710,7 @@ class SndRcvList(PacketList):
     __slots__ = []  # type: List[str]
 
     def __init__(self,
-                 res=None,  # type: Optional[Union[List[Packet], PacketList]]
+                 res=None,  # type: Optional[Union[List[Packet], PacketList, List[Tuple[Packet, Packet]]]]  # noqa: E501
```
Comment (guedou) [inline scapy/plist.py:713]: Why did you change the type of res`?

File: test/regression.uts [hunk 2]
```diff
@@ -13041,6 +13041,62 @@ p[IP]
 ############
 + PacketList methods
 
+= sr()
+
+class Req(Packet):
```
Comment (guedou) [inline test/regression.uts:13046]: Could you use an existing Packet instead? For example, `IP()/ICMP()ˋ?
I am ok for using these specific ones. I am just curious.

**Review by guedou — DISMISSED** [review]

**Comment by guedou** [comment]
Nice PR! Could you provide a use case?

**Comment by polybassa** [comment]
Hi,

On analysis of automotive protocols, I often have not answered messages. 

Every not answered message leads to a search over the full remaining list. This brings the runtime complexity nearly to O(n^2). Through the limitation of the lookahead, I can preserve the runtime complexity at O(k x n), if I can assume that I need to receive a answer within k packets. 
-- 
Cheers,

Nils

**Comment by polybassa** [comment]
In this line https://github.com/secdev/scapy/blob/bf2304bd534be00dd42f3c8ed3f605bb3c687b06/scapy/plist.py#L551

init of SndRcvList is called with a sr, which is a list of tuples. Therefore I modified the type.

Sorry for the custom packets. I wanted to have a defined behavior, and I don't know the internet based packets not so well. Also I didn't wanted to load the automotive contribution packets into regression. Do I need to change it?

**Comment by guedou** [comment]
Thanks for your answers. They helped me understand your changes. No need to modify anything.

**Comment by polybassa** [comment]
Great. Thanks.

**Comment by gpotter2** [comment]
LGTM but could we merge https://github.com/secdev/scapy/pull/2671 first ? It might raise some typing issues

**Comment by polybassa** [comment]
Sure. Thanks

-- 
Viele Grüße

Nils Weiß

**Comment by polybassa** [comment]
This PR is ready for merge.

**Review by gpotter2 — APPROVED** [review]
Thanks for reminding us :smile:
